KIPs/kip-237.md

```diff
@@ -0,0 +1,55 @@
+---
+kip: 237
+title: Storage Compression
+author: Lake (@hyunsooda), Ollie (@blukat29)
+discussions-to: TBD
+status: Draft
+category: Storage
+created: 2024-12-26
+---
+
+## Abstract
+Reduce disk storage by compressing data, cutting its size by 50% over.
+
+## Motivation
+Kaia has focused on optimizing disk storage, effectively reducing the state trie size through state migration and live pruning.
+However, these optimizations only address the state trie data. There is still significant potential for further reductions in storage.
+This KIP targets three types of data: header, body, and receipts.
+
+For each block processed, the header, body, and receipts data are generated.
+These data are progressively stored on disk, consuming a substantial amount of storage.
+Based on the migrated chain data (as of 12/02/2024), the header, body, and receipts account for 10.4%, 28.1%, and 40.9% of the total size, respectively, which together occupy 79.4% of the node's disk space.
+
+We propose implementing storage compression techniques to reduce the size of these three data types—header, body, and receipts—thereby improving disk space utilization.
+
+### Compression
+A node periodically compresses storage in a background thread and removes the original data once it is confirmed safe to do so.
+Data is compressed in units called "chunks," with the chunk size determined by a CLI flag.
+The default chunk size is 1MB. For example, if the target data size is 10MB, the data will be divided into 10 chunks.
+There is a tradeoff between speed and compression rate. Larger chunk sizes result in more effective compression but introduce higher compression and decompression overhead (time).
+
+### Decompression
+Decompression is required when a node receives a query (header, body, or receipts).
+The node first searches for the corresponding data in the original database.
+If the data is not found, it locates the decompressed chunk and performs the decompression. The decompressed chunk is cached for later use.
+
+### Retention
+Retention follows the same semantics as LivePruning, preserving the latest n blocks to prevent them from being compressed.
+
+### Evaluation
+We evaluated compression module on mainnet([chaindata of 12/02/2024](https://packages.kaia.io/mainnet/chaindata/)).
+| Item    | Header | Body | Receipts |
+| ------- | ------- | ------- | ------- |
+| Origin size | 410G | 1100G | 1600G |
+| Compressed size | 261G (40.24%) | 267G (75.73%) | 179G (88.81%) |
+
+Compression had been completed wihtin 3 days.
+
+### Ecosystem changes
```

Comment:
Maybe add a Rationale section with test results supporting the 1MB & 10,000 chunk size cap?

Reply:
Done https://github.com/kaiachain/kips/pull/37/commits/0f25bf76962154913ef34e480d6a0a304b6913c3